### PR TITLE
Fix a potential deadlock situation when shutting down or deleting a class

### PR DESCRIPTION
**Pull request by aliszka**

### CODE REVIEW

**Review by antas-marcin — APPROVED** [review]

**Review by parkerduckworth — APPROVED** [review]
This looks great

File: adapters/repos/db/lsmkv/bucket.go [hunk 1]
```diff
@@ -495,58 +486,48 @@ func (b *Bucket) Shutdown(ctx context.Context) error {
...
-				if shouldSwitch {
-					if err := b.FlushAndSwitch(); err != nil {
-						b.logger.WithField("action", "lsm_memtable_flush").
-							WithField("path", b.dir).
-							WithError(err).
-							Errorf("flush and switch failed")
-					}
-				}
-			}
+func (b *Bucket) initFlushCycle() {
```
Comment (etiennedi) [inline adapters/repos/db/lsmkv/bucket.go:489]: Does this name still make sense? If I see it correctly, the method no longer initiates a cycle, but rather makes a decision to flush if specific criteria are met. So maybe something along the lines of `maybeFlushAndSwitch` or `flushAndSwitchIfThresholdsMet`?

File: adapters/repos/db/lsmkv/segment_group.go [hunk 2]
```diff
@@ -292,22 +292,20 @@ func (sg *SegmentGroup) count() int {
...
-
-	select {
-	case <-ctx.Done():
+	if err := sg.compactionCycle.StopAndWait(ctx); err != nil {
 		return errors.Wrap(ctx.Err(), "long-running compaction in progress")
-	case <-compactionHalted:
 	}
 
+	// Lock acquirement placed after compaction cycle stop request, due to occasional deadlock,
+	// because compaction logic used in cycle also requires maintenance lock.
```
Comment (etiennedi) [inline adapters/repos/db/lsmkv/segment_group.go:300]: Great example of a comment outlining the motivation behind why a change was made. ❤️

File: adapters/repos/db/lsmkv/segment_group_compaction.go [hunk 3]
```diff
@@ -248,38 +247,21 @@ func (sg *SegmentGroup) stripTmpExtension(oldPath string) (string, error) {
 	return newPath, nil
 }
 
-func (sg *SegmentGroup) initCompactionCycle(interval time.Duration) {
-	if interval == 0 {
-		return
-	}
+func (sg *SegmentGroup) initCompactionCycle() {
```
Comment (etiennedi) [inline adapters/repos/db/lsmkv/segment_group_compaction.go:250]: Same as above with `initFlushCycle`, I think the `init` part no longer matches what this method does.

**Pull request merged**